### ISSUE TITLE
Make Completes Concurrent in Tx

### ISSUE THREAD

**Issue by abombss** [Particular/NServiceBus.AzureServiceBus#144]
https://github.com/Particular/NServiceBus.AzureServiceBus/blob/develop/src/Transport/Receiving/MessageReceiverNotifier.cs#L194

The completes can be merged and run concurrently when in a TX, no need to make them serialized, it adds latency.

**Comment by SeanFeldman** [comment]
@abombss sorry, not following. We're abandoning here, not completing.

**Comment by abombss** [comment]
Clipboard error :)
https://github.com/Particular/NServiceBus.AzureServiceBus/blob/develop/src/Transport/Receiving/MessageReceiverNotifier.cs#L161

**Comment by SeanFeldman** [comment]
@abombss I tried that before. 
```c#
var tasks = context.OnComplete.Select(toComplete => toComplete()).ToList();
tasks.Add(Complete(message));
await Task.WhenAll(tasks).ConfigureAwait(false);
```
But then completion is not working as expected ([test fails](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/develop/src/Tests/Topology/Operation/When_operating_StandardTopology.cs#L443)).

**Comment by yvesgoeleven** [comment]
@SeanFeldman @abombss  context.OnComplete is just a collection of tasks (belonging to methods that need to be finished before we complete the received message). Besides the timing, it has nothing to do with ASB's complete operation. So you should definitly not merge these .

**Comment by abombss** [comment]
But then I am creating lots more remote calls serialized correct?

Context.oncomplete is going to have any sends I do on the receive using via
correct? Then I send the receive complete, then the tx complete. That is 3
remote calls serialized instead of the needed 2? That can easily be an
additional 150 to 500 ms of additional latency.

**Comment by SeanFeldman** [comment]
@abombss ironically, what you pointed out was hiding another issue. When using OnMessageAPI, we need to autocomplete received messages (to allow prefetching). Therefore incoming message shouldn't have been completed manually by us [here](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/046a7a3476aebc89194e849dfe3a7c9154535dbd/src/Transport/Receiving/MessageReceiverNotifier.cs#L164).

**Comment by abombss** [comment]
Why would autocomplete impact prefetching? Isn't that going to impact the
ability to perform the complete in a tx along sends?

**Comment by SeanFeldman** [comment]
That's a good question. This is what our testing showed. Autocomplete set to true completes messages once processing callback is done (or abandons if exception is thrown). If Autocomplete is set to false and we autocomplete each received message individually, prefetch is rendered useless.
